Update variables documentation with latest APIs and better example (#1356)
fermyon/developer: content/spin/v2/variables.md

Spin supports dynamic application variables. Instead of being static, their values can be updated without modifying the application, creating a simpler experience for rotating secrets, updating API endpoints, and more. 

> These variables are defined in a Spin application manifest (in the `[variables]` section), and their values can be set or overridden at runtime by an [application variables provider](./dynamic-configuration.md#application-variables-runtime-configuration). When running Spin locally, the variables provider can be [Hashicorp Vault](./dynamic-configuration.md#vault-application-variable-provider) for secrets, or host environment variables.

Reviewer: Azure KeyVault is still there on main... are you saying this is being removed?  Is that part of the factors work?  Can we discuss?

Author: I originally started editing the `v1` file by accident and then copied it over. I clearly skipped over this v2 change that existed -- thank you!

For information about configuring application variables providers, refer to the [dynamic configuration documentation](./dynamic-configuration.md#application-variables-runtime-configuration).

## Adding Variables to Your Applications

Variables are added to an application under the top-level `[variables]` section of an application manifest (`spin.toml`). Each entry must either have a default value or be marked as `required = true`. “Required” entries must be [provided](./dynamic-configuration#application-variables-runtime-configuration) with a value.

> For example, say an application needs to access a bearer token for an outbound API call. A `[variables]` section could be added to an application's manifest with one entry for a variable named `token`. Since there is no reasonable default value for a secret, the variable is set as required with `required = true`. On the other hand, say the the API endpoint is known but should be configurable for A/B testing, another v`api_uri` variable could be added with a default that can be overridden. This scenario could be configured in the application manifest as follows:

Reviewer: ```suggestion
For example, say an application needs to access a bearer token for an outbound API call. A `[variables]` section could be added to an application's manifest with one entry for a variable named `token`. Since there is no reasonable default value for a secret, the variable is set as required with `required = true`. On the other hand, say the the API endpoint is known but should be configurable for A/B testing, another `api_uri` variable could be added with a default that can be overridden. This scenario could be configured in the application manifest as follows:
```

> For example, say an application needs to access a bearer token for an outbound API call. A `[variables]` section could be added to an application's manifest with one entry for a variable named `token`. Since there is no reasonable default value for a secret, the variable is set as required with `required = true`. On the other hand, say the the API endpoint is known but should be configurable for A/B testing, another v`api_uri` variable could be added with a default that can be overridden. This scenario could be configured in the application manifest as follows:

Reviewer: This feels like a lot of conditional mood ("could be added", "could be configured") - I wonder if we could make it read more positively and concretely.

```suggestion
For example, consider an application which makes an outbound API call using a bearer token. To configure this using variables, you would:

* Add a `[variables]` section to the application manifest
* Add a `token` entry for the bearer token.  Since there is no reasonable default value for this secret, set the variable as required with `required = true`.
* Add an `api_uri` variable.  The URL _is_ known, but is useful to override, for example for A/B testing. So you can give this variable a default value with `default = "http://my-api.com"`.

The resulting application manifest looks like this:
```

Maybe that is a bit _too_ definite though!

Author: I love it! You are a word artist.

<!-- @nocpy -->

```toml
[variables]
api_token = { required = true }
api_uri = { default = "http://my-api.com" }
```

Variables are surfaced to a specific component by adding a `[component.(name).variables]` section to the component and referencing them within it. The `[component.(name).variables]` section contains a mapping of component variables and values. Entries can be static (like `api_version` below) or reference an updatable application variable (like `token` below) using [mustache](https://mustache.github.io/)-inspired string templates. Only components that explicitly use variables in their configuration section will get access to them. This enables only exposing variables (such as secrets) to the desired components of an application.

<!-- @nocpy -->

```toml
[component.(name).variables]
token = "\{{ api_token }}"
api_uri = "\{{ api_uri }}"
api_version = "v1"
```

> Variables can also be surfaced to other sections of the application manifest that benefit from dynamic configuration. For example, the `allowed_outbound_hosts` can be dynamically configured using variables as follows:

Reviewer: ```suggestion
Variables can also be used in other sections of the application manifest that benefit from dynamic configuration. For example, the `allowed_outbound_hosts` can be dynamically configured using variables as follows:
```

<!-- @nocpy -->

```toml
[component.(name)]
allowed_outbound_hosts = [ "\{{ api_uri }}" ]
```

> When a component variable references an application variable, its value will dynamically update as the application variable changes. For example, if the `api_token` variable is provided using the [Spin Vault provider](./dynamic-configuration.md#vault-application-variable-provider), it can be updated by changing the value in HashiCorp Vault. The next time the component gets the value of `token`, the latest value of `api_token` will be returned by the provider. See the [next section](#using-variables-from-applications) to learn how to use Spin's configuration SDKs to get configuration variables within applications.

Reviewer: This refers specifically to component variables referencing application variables, so goes best either with the section on component variables (where it was before), or in a new section separated from the "variables in other parts of the manifest" stuff.

I think I'd prefer to move it back, and then we can add something here that variables in other parts of the manifest are substituted at application load time and not dynamically updated while the application is running.  Open to thoughts though!

Author: Good call. I agree that it fits better right after the component section.

{{ startTab "TypeScript"}}

> [**Want to go straight to the reference documentation?**  Find it here.](https://fermyon.github.io/spin-js-sdk/modules/Variables.html)

> Note that the name is `Config` rather than `Variables`.

```ts
import { HandleRequest, HttpRequest, HttpResponse, Config } from "@fermyon/spin-sdk"

export const handleRequest: HandleRequest = async function (request: HttpRequest): Promise<HttpResponse> {

>   let token = Config.get("token")

Reviewer: Heads up that this is changing in JS SDK v2 and there is #1355 in progress for that, so some synchronisation might be needed...

Author: Rebased!

> [**Want to go straight to the reference documentation?**  Find it here.](https://fermyon.github.io/spin-python-sdk/variables.html)

> The `variables` module has a function called `get`(https://fermyon.github.io/spin-python-sdk/variables.html#spin_sdk.variables.get).

Reviewer: Not your problem but I wonder why the link checker wasn't flagging the old broken link...

```bash
# Build the application
$ spin build
# Run the application, setting the values of the API token and URI via the environment variable provider
# using the `SPIN_VARIABLE` prefix (upper case is necessary as shown here)

> $ SPIN_VARIABLE_API_TOKEN="your-token-here" SPIN_VARIABLE_API_URI="http://my-best-api.com spin up

Reviewer: ```suggestion
$ SPIN_VARIABLE_API_TOKEN="your-token-here" SPIN_VARIABLE_API_URI="http://my-best-api.com" spin up
```
